src/lib/core/sio.c

```diff
@@ -240,12 +240,38 @@ sio_connect(int fd, const struct sockaddr *addr, socklen_t addrlen)
 	return rc;
 }
 
+static int
+sio_socketname_addr_to_buffer(int fd, const struct sockaddr *addr,
+			      socklen_t addrlen, char *buf, int size)
+{
+	int n = 0;
+	(void)n;
+	SNPRINT(n, snprintf, buf, size, "fd %d", fd);
+	SNPRINT(n, snprintf, buf, size, ", aka ");
+	SNPRINT(n, sio_addr_snprintf, buf, size, addr, addrlen);
+	return 0;
+}
+
 int
 sio_bind(int fd, const struct sockaddr *addr, socklen_t addrlen)
 {
 	int rc = bind(fd, addr, addrlen);
-	if (rc < 0)
-		diag_set(SocketError, sio_socketname(fd), "bind");
+	if (rc < 0) {
+		/* Preserve errno */
```

Comment:
The problem here is that it's major code duplication of the `sio_socketname` function. Let's try to find a better way to implement this.

E.g. we may make `sio_socketname_to_buffer` to accept `sockaddr`, which it'll use for its `aka ...`. At the same time we should rename `sio_socketname` to smth like `sio_socketname_addr` and make it accept `sockaddr` too. This way we can use `sio_socketname_addr` in our `sio_bind` and directly pass `sockaddr` we want.

At the end we'll introduce `sio_socketname` with the same API it has now. This func will get `sockaddr` from `getsocketname` and call `sio_scoketname_addr` with `fd` and this addr.

@LevKats what do you think?

Reply:
Interesting. But doesn't these 2 cases require different handling of `fd < 0`? I mean in old API address isn't printed for negative `fd` because I guess it's considered useless.  In `sio_bind` address should be printed anyway as we don't need to report actual state of `fd` but an attempt to bind this specific address. It seems like there may be a problem with printing peers too here:
```
rc = getpeername(fd, (struct sockaddr *) &addr, &addrlen);
if (rc == 0) {
	SNPRINT(n, snprintf, buf, size, ", peer of ");
	SNPRINT(n, sio_addr_snprintf, buf, size, base_addr, addrlen);
}
```
If you had `fd` already bound to some address it would fail to bind new one but it would show old peers here which may be counterintuitive.

So we may have 2 intents while using some new API: to print actual socket state and our attempt to bind address that doesn't seem to have much in common.

Reply:
> If you had fd already bound to some address it would fail to bind new one but it would show old peers here which may be counterintuitive.

You cannot have fd bound to some address, as this one is a newly created socket (in `evio_service_entry_bind_addr`):

https://github.com/tarantool/tarantool/blob/319357d5973d15d08b8eda6a230eada08b710802/src/lib/core/evio.c#L241-L258

So, no problems with peers.

>  But doesn't these 2 cases require different handling of fd < 0? I mean in old API address isn't printed for negative fd because I guess it's considered useless. In sio_bind address should be printed anyway as we don't need to report actual state of fd but an attempt to bind this specific address.

`fd < 0` means that we failed to create socket:

https://github.com/tarantool/tarantool/blob/319357d5973d15d08b8eda6a230eada08b710802/src/lib/core/sio.c#L172-L182

In such case we'll have errmsg: `binary: failed to bind on 8.8.8.8:3313: bind, called on fd -1`, which is also ok, I suppose